groups/bsl/bslstl/bslstl_optional.h

```diff
@@ -218,16 +219,243 @@ extern const Optional_OptNoSuchType optNoSuchType;
+struct Optional_AssignsFromStdOptional
+: bsl::integral_constant<
+      bool,
+      std::is_assignable<TYPE&, const std::optional<ANY_TYPE>&>::value ||
+          std::is_assignable<TYPE&, std::optional<ANY_TYPE>&>::value ||
+          std::is_assignable<TYPE&, const std::optional<ANY_TYPE> >::value ||
+          std::is_assignable<TYPE&, std::optional<ANY_TYPE> >::value> {
+};
+
+#else
+template <class TYPE, class ANY_TYPE>
+struct Optional_ConvertsFromStdOptional : bsl::integral_constant<bool, false> {
+};
+
+template <class TYPE, class ANY_TYPE>
+struct Optional_AssignsFromStdOptional : bsl::integral_constant<bool, false> {
+};
+#endif  // BSLS_LIBRARYFEATURES_HAS_CPP17_BASELINE_LIBRARY
+
+template <class TYPE, class ANY_TYPE>
+struct Optional_ConvertsFromOptional
+: bsl::integral_constant<
+      bool,
+      bsl::is_convertible<const bsl::optional<ANY_TYPE>&, TYPE>::value ||
+          bsl::is_convertible<bsl::optional<ANY_TYPE>&, TYPE>::value ||
+          bsl::is_convertible<const bsl::optional<ANY_TYPE>, TYPE>::value ||
+          bsl::is_convertible<bsl::optional<ANY_TYPE>, TYPE>::value
+              BSLSTL_OPTIONAL_OR_IS_CONSTRUCTIBLE_V(
+                  TYPE,
+                  const bsl::optional<ANY_TYPE>&)
+                  BSLSTL_OPTIONAL_OR_IS_CONSTRUCTIBLE_V(
+                      TYPE,
+                      bsl::optional<ANY_TYPE>&)
+                      BSLSTL_OPTIONAL_OR_IS_CONSTRUCTIBLE_V(
+                          TYPE,
+                          const bsl::optional<ANY_TYPE>)
+                          BSLSTL_OPTIONAL_OR_IS_CONSTRUCTIBLE_V(
+                              TYPE,
+                              bsl::optional<ANY_TYPE>) ||
+          Optional_ConvertsFromStdOptional<TYPE, ANY_TYPE>::value> {
+};
+
+#ifdef BSLS_LIBRARYFEATURES_HAS_CPP11_BASELINE_LIBRARY
+template <class TYPE, class ANY_TYPE>
+struct Optional_AssignsFromOptional
+: bsl::integral_constant<
+      bool,
+      std::is_assignable<TYPE&, const bsl::optional<ANY_TYPE>&>::value ||
+          std::is_assignable<TYPE&, bsl::optional<ANY_TYPE>&>::value ||
+          std::is_assignable<TYPE&, const bsl::optional<ANY_TYPE> >::value ||
+          std::is_assignable<TYPE&, bsl::optional<ANY_TYPE> >::value ||
+          Optional_AssignsFromStdOptional<TYPE, ANY_TYPE>::value> {
+};
+#else
+
+template <class TYPE, class ANY_TYPE>
+struct Optional_AssignsFromOptional : bsl::integral_constant<bool, false> {
+    //We only use '|| BloombergLP::bslstl::Optional_AssignsFrom' in
+    // 'bsl::optional' constraints. In order to ignore
+    // Optional_AssignsFromOptional trait in C++03, we set it to false so it
+    // never affects the trait it appears in.
+};
+#endif  //BSLS_LIBRARYFEATURES_HAS_CPP11_BASELINE_LIBRARY
+
+// Remove CV qualifiers and references from type 'U'
+#define BSLSTL_OPTIONAL_REMOVE_CVREF_T(U)                                     \
+    typename bsl::remove_cv<typename bsl::remove_reference<U>::type>::type
```

Comment:
Why a macro and not a struct?  "Simulating" `_t`? Also, this seems to be used only where types are expected, so even the outer `typename` disambugator is not required in `bsl::is_same<Optional_RemoveCVRef<T>::type, ....>`, right?

Also, this should ideally be `bslmf_removecvref` component, but perhaps this is out of scope.

Reply:
typename is required because whatever you put in bsl::is_same for this case will be a dependant name.
No specific reason to use macro, except that I already entered macro "land" with previous traits and it made it slightly shorter. :) I'll replace it with a trait. I can make a _t one, but I'd also like to add a _v to the Optional_IsConstructible and friends for some sort of symmetry - I hope that's fine with you.
As far as bslmf_removecvref goes, I agree that that would be the ideal solution, and yes, it I believe that is outside of scope for this SOW. I can ping Mike/Pablo and suggest they add it to the next SOW. There's quite a few 'minor' changes I've come across now that can probably be grouped into an SOW.

Reply:
Regarding the typename: I was basically referring to this working without need for `typename`: https://godbolt.org/z/7ce6x6

Reply:
that works because the template argument is not a dependant name there, but in all our uses of traits in 'optional', the type is a dependant name and the code does not compile without 'typename'. This is the case we have 
https://godbolt.org/z/31ef5T (note that you need 'typename' in both places in order to compile)

Reply:
Guess I was confused by the recent discussions of removing need for typename in more places for C++20.  Then I looked it up, found p0634 and looked at who the authors were... Thanks for putting up with me on this...

Reply:
lol :) I enjoy technical discussions. One's assumptions should regularly be challenged, and that includes my own assumptions (also, i found a leftover '#undef BSLSTL_OPTIONAL_REMOVE_CVREF_T' because of this comment).